python/cudf/cudf/core/tools/datetimes.py

```diff
@@ -0,0 +1,265 @@
+    >>> cudf.to_datetime(1490195805, unit='s')
+    numpy.datetime64('2017-03-22T15:16:45.000000000')
+    >>> cudf.to_datetime(1490195805433502912, unit='ns')
+    numpy.datetime64('1780-11-20T01:02:30.494253056')
+    """
+    if arg is None:
+        return None
+
+    if exact is False:
+        raise NotImplementedError("exact support is not yet implemented")
+
+    if origin != "unix":
+        raise NotImplementedError("origin support is not yet implemented")
+
+    if yearfirst:
+        raise NotImplementedError("yearfirst support is not yet implemented")
+
+    try:
+        if isinstance(arg, cudf.DataFrame):
+            # we require at least Ymd
+            required = ["year", "month", "day"]
+            req = sorted(list(set(required) - set(arg._data.names)))
+            if len(req):
+                raise ValueError(
+                    "to assemble mappings requires at least that "
+                    "[year, month, day] be specified: [{required}] "
+                    "is missing".format(required=",".join(req))
+                )
+
+                # replace passed unit with _unit_map
+
+            def get_units(value):
+                if value in _unit_map:
+                    return _unit_map[value]
+
+                # m is case significant
+                if value.lower() in _unit_map:
+                    return _unit_map[value.lower()]
+
+                return value
+
+            unit = {k: get_units(k) for k in arg._data.names}
+            unit_rev = {v: k for k, v in unit.items()}
+
+            # keys we don't recognize
+            excess = sorted(set(unit_rev.keys()) - set(_unit_map.values()))
+            if len(excess):
+                excess = ",".join(excess)
+                raise ValueError(
+                    f"extra keys have been passed to the \
+                        datetime assemblage: [{excess}]"
+                )
+
+            new_series = (
+                arg[unit_rev["year"]].astype("str")
+                + "-"
+                + arg[unit_rev["month"]].astype("str").str.zfill(2)
+                + "-"
+                + arg[unit_rev["day"]].astype("str").str.zfill(2)
+            )
+            format = "%Y-%m-%d"
+            col = new_series._column.as_datetime_column(
+                "datetime64[ns]", format=format
+            )
+
+            time_delta_col = None
+            for u in ["h", "m", "s", "ms", "us", "ns"]:
+                value = unit_rev.get(u)
+                if value is not None and value in arg:
+                    current_col = arg._data[value]
+                    # If the arg[value] is of int or
+                    # float dtype we don't want to type-cast
+                    if current_col.dtype.kind not in ("i", "f"):
+                        current_col = current_col.astype(dtype="int64")
+
+                    factor = utils.scalar_broadcast_to(
+                        scalar=column.datetime._numpy_to_pandas_conversion[u],
```

Comment:
We don't necessarily want to convert to `ns` similar to Pandas. We have `s`, `ms`, `us`, `ns` implementations where we can avoid conversion in many situations, and only have to convert to `s` in the case of `h` / `m`.

Reply:
Reworked on casting and binary op on `factor`.
Made changes so that we do not go down till `ns` for calculations, but there are cases like in case of float inputs we will have to cast down till `ns` for calculations as the higher the number of decimal values in floating number, the higher is chance of calculations need to spill over to `ns` level.